### DISCUSSION TITLE
How to keep eslint errors from failing a development build

### DISCUSSION THREAD

**Discussion by sallf** [nrwl/nx#10995] [General]
I just finished copying a big project into nx with a ton of linter errors. I want to continue to see them in my IDE, but I don't want them to cause the build to fail when running `nx serve <my-app>`. When running that I see `webpack compiled successfully` but then get multiple errors that reference non-breaking linter errors.

I've tried adding a custom `webpack.config` to my app and ensuring that `failOnError: false` is passed in as an option, but still I'm getting these errors. 

Any help greatly appreciated.

**Comment by rfprod** [comment]
@sallf Are you sure it's linter errors, maybe it is compiler? Linter errors do not cause build/serve to fail, they cause linting commands like `npx nx lint client` to fail.
![Screenshot from 2022-07-03 11-41-26](https://user-images.githubusercontent.com/1507770/177032352-442a31ff-7639-408b-8283-96c001965ac0.png)

**Reply by sallf** [reply]
@rfprod I think it successfully compiles. So from `nx serve <my-project>` I see this:

<img width="484" alt="Screen Shot 2022-07-03 at 10 45 05 AM" src="https://user-images.githubusercontent.com/5176792/177049509-d2dffb0f-9b40-499b-9834-5cd6e52f6877.png">

Here's the same error in VSCode. Again, shouldn't affect compiling, I just haven't defined a type for the arg, which is not a breaking error.

![Screen Shot 2022-07-03 at 10 49 31 AM](https://user-images.githubusercontent.com/5176792/177049566-30eb0630-784b-4da5-8f3b-ee241f057226.png)

And then here's what it's showing in the browser. 

<img width="1492" alt="Screen Shot 2022-07-03 at 10 46 38 AM" src="https://user-images.githubusercontent.com/5176792/177049672-e7cf3c3f-31e6-46d7-9961-316dc853ff15.png">

I'm at a total loss as to why I can't seem to ignore these.

**Edit**
I guess it is failing to compile (since it says "Failed to compile" at the top of the browser). Maybe I don't understand why it's failing. Everything ran fine in the old project location which was also using webpack.

**Reply by rfprod** [reply]
@sallf it is a TypeScript compiler error, definitely not a linter error. Could you please share the whole code snippet so that it's visible?

Does the previous project location use TypeScript as well? It seems like you move JavaScript sources to TypeScript codebase. If so, it won't work like this. You have to have correct inferred types, or if type inference does not work, you will have to be explicit about types. Generally, type inference works well with well organized code bases and clean code.
In this case that you shared TypeScript compiler thinks `el` type is `never` instead of `HTMLElement` or a specific element like `HTMLDivElement`.

An example of TypeScript code (strict mode + null checks + explicit types) that compiles

```typescript
const findTabbableDescendants = (modal: HTMLElement): HTMLElement[] => {
   const result: HTMLElement[] = [];
   // your function logic
   return result;
}

const focusModal = () => {
  const modal: HTMLElement | null = document.getElementById('modal') // I don't see full source code in your screenshot, just guessing
  if (modal !== null) {
    const descendants: HTMLElement[] = findTabbableDescendants(modal); // what does this function return? possibly an array with DOM elements or null or an empty array. I'll assume it's an array that may be empty.
    const el: HTMLElement = descendants[0] ?? modal;
    el.focus();
  }
}
```

**Reply by sallf** [reply]
@rfprod the previous project was using TypeScript. Changing `strict` to `false` definitely helped remove some errors. And thanks for the example, I really appreciate your support. I think I know how to fix most of the errors, I was just really hoping to do it later. 

I guess in the previous project location we were using the linter to see TypeScript errors in VS Code, but then maybe weren't using it to compile the code? I don't know if that even makes sense, I'm at a total loss as to how this was working in the previous location and is breaking now. I tried copying the previous webpack config into the new location, and that didn't seem to help either. We were using `babel`. This is what the `ts` portion looked like in the old location:

```
rules: [
  {
    test: /\.(ts|js)x?$/,
    include: [
      path.resolve(__dirname, 'src'),
      path.resolve(__dirname, 'tests'),
    ],
    exclude: /node_modules/,
    use: [
      {
        loader: 'babel-loader',
        options: {
          cacheDirectory: true,
          presets: [
            "@babel/preset-env",
            "@babel/preset-react",
            "@babel/preset-typescript",
          ],
        },
      },
      {
        loader: 'source-map-loader',
      },
    ],
  },
  ...
]
```

And the `tsconfig` from the previous location:

```
{
  "compilerOptions": {
    "baseUrl": ".",
    "sourceMap": true,
    "paths": {
      "$common/*": ["./src/common/*"],
      "$components/*": ["./src/components/*"],
      "$constants/*": ["./src/constants/*"],
      "$features/*": ["./src/redux/features/*"]
    },
    "plugins": [{ "name": "typescript-plugin-css-modules" }],
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "strict": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react",
    "target": "es2016", 
    "module": "commonjs",
    "composite": true
  },
  "include": ["src"],
  "exclude": ["node_modules"]
 }

```

Anyways, at this point I think I'll just fix the errors, which is maybe what I should have done 15 hours ago.

**Comment by sallf** [comment]
I don't know if it's helpful, but here's what my `nx.json` looks like:

```
{
  "$schema": "./node_modules/nx/schemas/nx-schema.json",
  "npmScope": "nx-test",
  "affected": {
    "defaultBase": "main"
  },
  "implicitDependencies": {
    "package.json": {
      "dependencies": "*",
      "devDependencies": "*"
    },
    ".eslintrc.json": "*",
    "tools/eslint-rules/**/*": "*"
  },
  "tasksRunnerOptions": {
    "default": {
      "runner": "nx/tasks-runners/default",
      "options": {
        "cacheableOperations": ["build", "lint", "test", "e2e"]
      }
    }
  },
  "targetDefaults": {
    "build": {
      "dependsOn": ["^build"]
    }
  },
  "generators": {
    "@nrwl/react": {
      "application": {
        "style": "scss",
        "linter": "eslint",
        "babel": true
      },
      "component": {
        "style": "scss"
      },
      "library": {
        "style": "scss",
        "linter": "eslint"
      }
    }
  },
  "defaultProject": "nx-text"
}
```

And the `serve` target inside my `project.json` in the app I'm trying to build.

```
"serve": {
      "executor": "@nrwl/web:dev-server",
      "defaultConfiguration": "development",
      "options": {
        "buildTarget": "kap-stake:build",
        "hmr": true
      },
      "configurations": {
        "development": {
          "buildTarget": "kap-stake:build:development"
        },
        "production": {
          "buildTarget": "kap-stake:build:production",
          "hmr": false
        }
      }
    },
```

**Reply by rfprod** [reply]
@sallf what about your root `tsconfig.json` or `tsconfig.base.json`?

**Reply by sallf** [reply]
@rfprod here's `tsconfig.base.json`:

```
{
  "compileOnSave": false,
  "compilerOptions": {
    "rootDir": ".",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "target": "es2015",
    "module": "esnext",
    "lib": ["es2017", "dom"],
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "baseUrl": ".",
    "noPropertyAccessFromIndexSignature": false,
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "jsx": "react",
    "paths": {
      "@nx-test/kap-stake-components/*": ["libs/kap-stake/components/src/*"],
      "@nx-test/kap-stake-screens/*": ["libs/kap-stake/screens/src/*"],
      "@nx-test/kap-stake/constants": ["libs/kap-stake/constants/src/index.ts"],
      "@nx-test/kap-stake/helpers": ["libs/kap-stake/helpers/src/index.ts"]
    }
  },
  "exclude": ["node_modules", "tmp"]
}
```

... and `tsconfig.json` for the app:

```
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "jsx": "react-jsx",
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
  },
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.app.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ]
}
```

The `tsconfig.app.json` and `tsconfig.spec.json` should be 100% from the generator (`npx nx generate @nrwl/react:application my-application`).

**Reply by rfprod** [reply]
@sallf I would recommend that you address all compiler errors rather than modifying `tsconfig.json`. See this comment https://github.com/nrwl/nx/discussions/10995#discussioncomment-3074098
Otherwise, if you consider runtime errors and unexpected app behavior acceptable, first of all try turning off the [strict mode](https://www.typescriptlang.org/tsconfig#strict).
`tsconfig.json`
```json
...
"compilerOptions": {
    "jsx": "react-jsx",
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "strict": false, <--
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
},
...
```
